manual/CHAPTER_TextRtlil.tex

```
@@ -0,0 +1,270 @@
+
+\section{File}
+
+A file consists of zero or more modules, attribute statements, and auto-index statements. All statements terminate in an end-of-line. Because of this, statements cannot contain end-of-lines.
+
+Attributes at the file level are applied to the following module.
+
+\begin{indentgrammar}{<design>}
+<file> ::= (<module> | <attr-stmt> | <autoidx-stmt>)$*$
+\end{indentgrammar}
+
+\subsection{Modules}
+
+Declares a module consisting of zero or more attributes, wires, memories, cells, processes, and connections.
+
+\begin{indentgrammar}{<module-body-stmt>}
+<module> ::= <module-stmt> <module-body> <module-end-stmt>
+
+<module-stmt> ::= "module" <id> <eol>
+
+<module-body> ::= 
+(<param-stmt>
+  \alt <attr-stmt>
+  \alt <wire-stmt>
+  \alt <memory-stmt>
+  \alt <cell>
+  \alt <process>
+  \alt <conn-stmt>)$*$
+\end{indentgrammar}
+
+<module-end-stmt> ::= "end" <eol>
+
+\subsection{Signal specifications}
+
+A signal is anything that can be applied to a cell port, i.e. a constant value, all bits or a selection of bits from a wire, or concatenations of those.
+
+See Sec.~\ref{sec:rtlil_sigspec} for an overview of signal specifications.
+
+\begin{indentgrammar}{<sigspec>}
+<sigspec> ::=
+<constant>
+    \alt <wire-id>
+    \alt <sigspec> "[" <integer> (":" <integer>)$?$ "]"
+    \alt "\{" <sigspec>$*$ "\}"
+\end{indentgrammar}
+
+\subsection{Connection statements}
+
+Declares a connection between the given signals.
+
+\begin{indentgrammar}{<conn-stmt>}
+<conn-stmt> ::= "connect" <sigspec> <sigspec> <eol>
+\end{indentgrammar}
+
+\subsection{Attribute statements}
+
+Declares an attribute with the given identifier and value. Attributes at the file level apply to the following module. Attributes within a module apply to the following non-attribute statement.
+
+\begin{indentgrammar}{<attr-stmt>}
+<attr-stmt> ::= "attribute" <id> <constant> <eol>
+
+<constant> ::= <value> | <integer> | <string>
```

It is worth noting that plain integer constants have a really weird, unintuitive behavior: whatever integer you have in the text is silently truncated to 32 bits and converted to unsigned. I had a PR to improve this (https://github.com/YosysHQ/yosys/pull/2066) but it got rejected in favor of a different approach... one we have yet to determine.

---

Noted in document.